### ISSUE TITLE
xdotool key/type breaks with multiple keyboards with different layouts

### ISSUE THREAD

**Issue by m-ou-se** [jordansissel/xdotool#150]
I have two keyboards with different layouts. Depending on the last one I used, xdotool works or doesn't work.

`xdotool key apostrophe` produces `'` after using my US layout keyboard, and produces `p` on my custom layout keyboard. `'` can be typed as `Mod5+p` on this keyboard.

After a bit of testing, I believe the following is happening: (I don't know much about X or xdotool, though, so I might be wrong.) xdotool looks up the key (combination) used to make `'` **using the current layout of the currently active keyboard** (`Mod5+p`), but then executes it using the **XTEST** keyboard, which does not use the custom layout, and doesn't even have a key for `Mod5`. Only the `p` keypress goes through. (And even if it managed to press `Mod5`, it;d still not produce an `'`, since `Mod5+p` doesn't have a special meaning in the layout the XTEST keyboard uses.)

I changed the custom layout to put `slash` on the keycode that produces `grave` on the layout that my other keyboard and XTEST use.  `xdotool key slash` then produces `` ` `` when I press enter on my custom layout keyboard, but produces `/` when I press enter using my US layout keyboard.

To confirm, I applied my custom keyboard layout to the XTEST keyboard (using `xkbcomp -i`), and afterwards it worked fine. (Surpsiringly, then it also worked when US layout keyboard was active. I'm not sure how that works.)

**Comment by jordansissel** [comment]
I think your assessment of the bug is correct. Keyboards are sooo complex in X11! hah :\

**Comment by jschwartzenberg** [comment]
I can confirm this bug. I have two layouts set up, us and ru. When I try to do either this with US as active:
`xdotool key colon`
or
`xdotool send :`

It ends up either typing a 6 or a ^ instead. Even when my active layout is US. Shift+6 produces a colon with the Russian layout.

If anybody knows a workaround, that would be really great!

**Comment by jschwartzenberg** [comment]
Similarly with other keys:
`xdotool key semicolon` types a 4
`xdotool key shift+semicolon` types a $.

**Comment by m-ou-se** [comment]
@jschwartzenberg A dirty workaround is to first let xdotool type something, anything at all, to switch the active keyboard to the one xdotool uses (XTEST). Then invoke xdotool again, without touching your keyboard, so XTEST remains active while it does its lookup. The second one will work correctly.

For example: `xdotool key space && xdotool type :`

Maybe you can use a key that has no effect, instead of space. (F20, or Cancel, or something like that.) Or just send a backspace after the space. ;)

**Comment by mkoskar** [comment]
It seems that applying same settings to both "core keyboard" and "XTEST keyboard" is a workaround. I suppose this wouldn't work well in case of per-keyboard settings as @jschwartzenberg described.

E.g., I do this:
```
opts=(                                                                    
    -model pc105                                                          
    -layout us,sk,cz                                                      
    -variant cz_sk_de,qwerty,qwerty                                       
    -option                                                               
    -option caps:ctrl_modifier                                            
    -option keypad:pointerkeys                                            
    -option lv3:ralt_switch_multikey                                      
    -option nbsp:level3                                                   
    -option shift:both_capslock_cancel                                    
    -option terminate:ctrl_alt_bksp                                       
)                                                                         
                                                                          
setxkbmap "${opts[@]}"                                                    
                                                                          
xtest_kbd=$(xinput list --id-only 'keyboard:Virtual core XTEST keyboard') 
[[ $xtest_kbd ]] && setxkbmap "${opts[@]}" -device "$xtest_kbd"
```

**Comment by DreadPirateLynx** [comment]
I'd just like to say that this is still a problem that needs to be addressed. I currently use xkb to program my Razer Tartarus with application specific profiles, and xbindkeys to program my mouse. The workaround listed above (xdotool key F20 && xdotool key Return) works well enough for "normal" keys, but modifiers are all but impossible; something always eventually ends up getting stuck.

**Comment by jordansissel** [comment]
I agree it'd be nice to solve. I spent some time years ago trying to fix whatever causes this problem where xdotool makes mistakes somehow in what it learns about the keyboard mapping, but I was never successful in fixing it. :\

**Comment by DreadPirateLynx** [comment]
If it's any consolation, xte has similar problems

**Comment by blueyed** [comment]
For me the following workaround works: `xdotool key --clearmodifiers shift && xdotool type '…'`, as per @m-ou-se's comment (https://github.com/jordansissel/xdotool/issues/150#issuecomment-284409367).
The Shift is a no-op to make the XTEST keyboard active first.

**Comment by jordansissel** [comment]
Oh that is interesting! I wonder if we could build this workaround into
xdotool?
